keel-core/src/main/kotlin/com/netflix/spinnaker/keel/state/StateInspector.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2017 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.keel.state
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.keel.annotation.Computed
+import org.slf4j.LoggerFactory
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.primaryConstructor
+
+data class FieldState(
+  val name: String,
+  val current: Any?,
+  val desired: Any?
+) {
+  fun changed() = current != desired
+}
+
+/**
+ * Allows callback mutations of individual fields before building FieldState objects.
+ */
+data class FieldMutator(
+  val name: String,
+  val mutator: (v: Any?) -> Any?
+)
+
+/**
+ * Performs basic state diffing to determine whether or not system state needs to be converged on the desired spec.
+ */
+class StateInspector(
+  private val objectMapper: ObjectMapper
+) {
```

Comment:
If we find value, we'll also be able to open up an API offering granular diffs of resource vs desired state.